service/round.go

```diff
@@ -115,50 +164,107 @@ func newRound(dbdir, datadir string, epoch uint32, maxMembers uint) (*round, err
+		flushInterval:  opts.submitFlushInterval,
 	}
 
 	membersCounter.Add(float64(r.members))
 
 	return r, nil
 }
 
-func (r *round) submit(ctx context.Context, key, challenge []byte) error {
+// submit a challenge to the round under the given key.
+// The challenges are collected in a batch and persisted to disk periodically.
+// Returns an error if it was not possible to add the challenge to the batch (for example, duplicated key)
+// and a channel to which the final result will be sent when the challenge is persisted.
+// The caller must await the returned channel to make sure that the challenge is persisted.
+func (r *round) submit(ctx context.Context, key, challenge []byte) (<-chan error, error) {
 	if !r.isOpen() {
-		return ErrRoundIsNotOpen
+		return nil, ErrRoundIsNotOpen
 	}
 
-	if r.members >= r.maxMembers {
-		return ErrMaxMembersReached
+	r.batchMutex.Lock()
+	defer r.batchMutex.Unlock()
+
+	if r.members+uint(len(r.pendingSubmits)) >= r.maxMembers {
+		return nil, ErrMaxMembersReached
+	}
+
+	if r.batch == nil {
+		r.pendingSubmits = make(map[string]pendingSubmit)
+		r.batch = leveldb.MakeBatch(r.maxBatchSize)
+		time.AfterFunc(r.flushInterval, r.flushPendingSubmits)
+	}
+
+	pending, ok := r.pendingSubmits[string(key)]
+	if ok {
+		switch {
+		case bytes.Equal(challenge, pending.challenge):
+			return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
+		default:
+			return nil, ErrConflictingRegistration
+		}
 	}
 
 	registered, err := r.challengesDb.Get(key, nil)
 	switch {
 	case errors.Is(err, leveldb.ErrNotFound):
+
 		// OK - challenge is not registered yet.
 	case err != nil:
-		return fmt.Errorf("failed to check if challenge is registered: %w", err)
+		return nil, fmt.Errorf("failed to check if challenge is registered: %w", err)
 	case bytes.Equal(challenge, registered):
-		return fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
+		return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
 	default:
-		return ErrConflictingRegistration
+		return nil, ErrConflictingRegistration
 	}
 
-	if err := r.challengesDb.Put(key, challenge, &opt.WriteOptions{Sync: true}); err != nil {
-		return fmt.Errorf("failed to register challenge: %w", err)
+	r.batch.Put(key, challenge)
+	done := make(chan error, 1)
+	r.pendingSubmits[string(key)] = pendingSubmit{
+		done,
+		challenge,
 	}
 
-	r.members += 1
-	r.membersCounter.Inc()
+	if r.batch.Len() >= r.maxBatchSize {
+		r.flushPendingSubmitsLocked()
+	}
 
-	return nil
+	return done, nil
+}
+
+func (r *round) flushPendingSubmits() {
+	r.batchMutex.Lock()
+	defer r.batchMutex.Unlock()
+	r.flushPendingSubmitsLocked()
+}
+
+func (r *round) flushPendingSubmitsLocked() {
+	if r.batch == nil {
+		return
+	}
+	logging.FromContext(context.Background()).
+		Debug("flushing pending submits", zap.Int("num", len(r.pendingSubmits)), zap.String("round", r.ID))
+	err := r.challengesDb.Write(r.batch, &opt.WriteOptions{Sync: true})
```

Comment:
histogram with write latency will be useful to adjust max size and wait time

Reply:
Added an exponential bucket from 0.00001 with step 2x and 20 buckets. Theoretically, the NVME drives on the BM should be capable of around 4000ns latency but include up to the higher 5s range to detect misbehavior. Wdyt?
```
poet_round_batch_write_latency_seconds_bucket{le="1e-05"} 0
poet_round_batch_write_latency_seconds_bucket{le="2e-05"} 0
poet_round_batch_write_latency_seconds_bucket{le="4e-05"} 0
poet_round_batch_write_latency_seconds_bucket{le="8e-05"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00016"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00032"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00064"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00128"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00256"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.00512"} 0
poet_round_batch_write_latency_seconds_bucket{le="0.01024"} 71
poet_round_batch_write_latency_seconds_bucket{le="0.02048"} 144
poet_round_batch_write_latency_seconds_bucket{le="0.04096"} 170
poet_round_batch_write_latency_seconds_bucket{le="0.08192"} 170
poet_round_batch_write_latency_seconds_bucket{le="0.16384"} 170
poet_round_batch_write_latency_seconds_bucket{le="0.32768"} 171
poet_round_batch_write_latency_seconds_bucket{le="0.65536"} 171
poet_round_batch_write_latency_seconds_bucket{le="1.31072"} 171
poet_round_batch_write_latency_seconds_bucket{le="2.62144"} 171
poet_round_batch_write_latency_seconds_bucket{le="5.24288"} 171
```